File: JavaScript/0006.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function(s, numRows) {
    // Edge cases
    if (numRows === 1 || numRows >= s.length) {
        return s;
    }
    
    // Initialize an array of empty strings for each row
    let rows = new Array(numRows).fill('').map(() => '');
    let currentRow = 0;
    let goingDown = false;
    
    // Traverse each character in the string
    for (let i = 0; i < s.length; i++) {
        rows[currentRow] += s[i];
        
        // Change direction if we're at the top or bottom row
        if (currentRow === 0 || currentRow === numRows - 1) {
            goingDown = !goingDown;
        }
        
        // Move to the next row
        currentRow += goingDown ? 1 : -1;
    }
    
    // Concatenate all rows to get the final result
    return rows.join('');
};

File: Ruby/0006.rb
# @param {String} s
# @param {Integer} num_rows
# @return {String}
def convert(s, num_rows)
    return s if num_rows == 1 || num_rows >= s.length
  
    # Initialize an array of empty strings for each row
    rows = Array.new(num_rows, '')
  
    current_row = 0
    going_down = false
  
    # Traverse each character in the string
    s.each_char do |char|
      rows[current_row] += char
  
      # Change direction if we're at the top or bottom row
      if current_row == 0 || current_row == num_rows - 1
        going_down = !going_down
      end
  
      # Move to the next row
      current_row += going_down ? 1 : -1
    end
  
    # Concatenate all rows to get the final result
    rows.join('')
  end
  